### ISSUE TITLE
P2: Boundary Loss (Kervadec et al., 2019) — differentiable surface distance proxy

### ISSUE THREAD

**Issue by petteriTeikari** [petteriTeikari/vascadia#100]
## Summary

Implement **Boundary Loss** from Kervadec et al. (2019) as a differentiable proxy for surface distance supervision during training. This is the recommended approach from our compound loss research for adding MASD-like optimization signal.

**Paper**: Kervadec, Bouchtiba, Desrosiers, Granger, Dolz, and Ben Ayed, "Boundary loss for highly unbalanced segmentation," MIDL 2019 (runner-up best paper), extended in Medical Image Analysis, 2021.
**arXiv**: https://arxiv.org/abs/1812.07032
**Official repo**: https://github.com/LIVIAETS/boundary-loss

## How It Works

Instead of computing integrals over segmentation *regions* (like Dice), Boundary Loss computes integrals over the *boundary*:

```
L_boundary = sum(softmax_pred * signed_distance_map)
```

- Signed distance maps (SDT) are **precomputed** at the dataloader level using `scipy.ndimage.distance_transform_edt`
- Negative inside the object, positive outside
- Combined with Generalized Dice: `L_total = L_GDL + 0.01 * L_boundary`

## Why This Matters for MiniVess

1. **MASD is our primary compound metric** (`val_compound_masd_cldice`) but MASD itself is non-differentiable (hard thresholding, surface extraction, argmin — see compound-loss-implementation-plan.md Section 2.2)
2. Boundary Loss is the closest differentiable proxy for average surface distance
3. v2 training results show a clear **topology-accuracy tradeoff**: `dice_ce` wins DSC/MASD while `dice_ce_cldice` wins clDice — a boundary loss could help bridge this gap

## Implementation Path

1. **Dataloader modification**: Precompute signed EDT for each volume during `CacheDataset` caching (one-time cost)
2. **Loss class**: `BoundaryLoss(nn.Module)` — simple `sum(softmax * sdt)` operation
3. **Compound losses**: `dice_ce + boundary` and `dice_ce_cldice + boundary` (triple compound: region + topology + surface)
4. **Config**: Add to `dynunet_losses.yaml` for v3 ablation

## Properties

| Property | Value |
|----------|-------|
| **Patch-safe** | Yes (precomputed EDT travels with patches) |
| **VRAM overhead** | ~10% (one extra channel for SDT) |
| **In MONAI** | No (milestone exists but not implemented) |
| **Proven for vessels** | Tested on brain lesions (ISLES, WMH), not vessel-specific |

## Known Limitations

- **Unbounded output**: Distance values range from -inf to +inf, requires careful weighting (alpha=0.01 in original paper)
- **Anisotropic spacing**: SDT computation must account for variable voxel spacing per axis
- **Not in MONAI**: Requires custom implementation (reference: official repo above)

### Planning Documents
- `docs/planning/compound-loss-implementation-plan.md` — Section 2.3: identifies Boundary Loss as recommended surface proxy
- `docs/planning/loss-and-metrics-double-check-report.md` — Section 3.3: patch-safety analysis, Section 6.2: P1 priority for v3
- `docs/planning/loss-metric-improvement-implementation.xml` — v2 plan (Boundary Loss deferred to v3)
- `docs/planning/experiment-planning-and-metrics-prompt.md` — original loss/metric requirements
- `docs/planning/topology-loss-plan.md` — topology loss research context
- `docs/planning/multi-metric-downstream-double-check.md` — downstream artifact gaps

### Related Issues
- #85 (closed): CbDiceClDiceLoss compound loss (current topology approach)
- #86 (closed): MetricsReloaded validation metrics during training (MASD tracking)
- #87 (closed): Experiment config v2 with compound primary metric
- #5 (closed): Topology-aware loss functions (clDice, cbDice, Betti)
- See also: Generalized Surface Loss issue (companion to this one)

### Key Files
- `src/minivess/pipeline/loss_functions.py` — loss factory (add new entry)
- `src/minivess/pipeline/validation_metrics.py` — MASD normalization (evaluation counterpart)
- `configs/experiments/dynunet_losses.yaml` — experiment config (add v3 losses)

## References

- Kervadec et al. (2019) — "Boundary loss for highly unbalanced segmentation" (MIDL, runner-up best paper)
- Karimi and Salcudean (2019) — "Reducing the Hausdorff Distance in Medical Image Segmentation with CNNs" (IEEE TMI)
- Shit et al. (2021) — clDice (CVPR) — topology counterpart to boundary supervision
